blog/content/post/2023-06-07-twpm-spi-fix.md

```diff
@@ -0,0 +1,855 @@
+---
+title: Using SPI slave on STM32L476 platform.
+abstract: 'Trustworthy Platform Module - the open-source and trustworthy TPM. A
+          story about implementing SPI interface for TwPM firmware.'
+cover: /img/nucleo-l476rg.jpg
+author: artur.kowalski
+layout: post
+published: true
+date: 2023-06-07
+archives: "2023"
+
+tags:
+  - embedded
+  - spi
+  - Zephyr
+  - STM32
+  - STMicroelectronics
+  - TPM
+  - TwPM
+categories:
+  - Firmware
+  - Miscellaneous
+  - Security
+
+---
+
+STM32 MCUs come with various peripherals, one of them is SPI (Serial Peripheral
+Interface) which is a simple serial bus interface commonly used for
+short-distance communication between various devices. SPI is one of the
+interfaces used by TPM chips for communication with PC motherboard. SPI uses 4
+lines for communication: MOSI, MISO, SCK, SS, which are described down below.
+
+The device must implement the TPM protocol to work as a TPM module. TPM protocol
+works by transmitting variable-length frames over SPI. The 4-byte TPM header
+contains fields describing the length of the data payload, the address of the
+target TPM register, and the transfer direction (read or write). TPM  protocol
+has its own means of handling flow control (as there isn't a standard flow
+control mechanism on SPI) and for doing bus aborts.
+
+TPMs typically operate at frequencies from 10 MHz up to 24 MHz. TPM must be able
```

Comment:
@macpijan I reworked this section a bit to avoid making assumptions. Added sources, also I found out that PTP spec encourages support for 33-66 MHz frequency range. That's however not required by the spec, the required frequency range is still 10-24 MHz.